### DISCUSSION TITLE
CRUD operations in offline mode - Firestore - Android

### DISCUSSION THREAD

**Discussion by AkshayAdiga1997** [invertase/react-native-firebase#5244] [General]
If I have a where clause based on modified_date (firestore servertimestamp) field 
`where('md', '>=', last_fetch_firestore_timestamp)`
, I get the document in IOS even when offline. The metadata.hasPendingWrites of the document would be true (even though the modified_date field value is now null) But the same scenario fails on android.
Note: where clause is required. Pl let me know what am I missing here. Thanks

**Comment by mikehardy** [comment]
This may be due to underlying platform differences - I strongly suspect it is actually.
If you've checked the actual code through here to make sure that the underlying javascript (then Obj-C and Java) API calls seem sensible, my best - though unfortunately difficult - suggestion is to make a quickstart for each platform where you verify behavior and confirm whether it's an upstream issue or not

https://github.com/firebase/quickstart-ios/tree/master/firestore
https://github.com/firebase/quickstart-android/tree/master/firestore

I strongly recommend checking both of those repo's issues list (especially android since that appears to be the platform not meeting expectations?) to see if they have anything.

Also make absolutely sure you are on the most up to date react-native-firebase version so you have the updated native SDKs

**Reply by AkshayAdiga1997** [reply]
First of all, thanks for this great library!

Yes. I am currently on version 11.4.1 (which is the latest as of now). This is the scenario.
I want to fetch data from a collection from firestore. But each time I fetch, I want to fetch only the delta (added/updated) documents. This is my approach.

I have a field called **modified_date** (shorthand md) and **local_modified_date** (shorthand lmd). The modified_date has the value `firestore.FieldValue.serverTimestamp()` and local_modified_date has the value `firestore.Timestamp.now()`. These values are computed while adding/updating a document.

So first I fetch all the documents from my local cache and store it in redux. Meanwhile I find out the modified_date of the latest document (by comparing modified_data.seconds value). I convert it into date format. Let's call it **last_fetch_timestamp**.

Now I read data from server where md (modified_date) is greater than equal to last_fetch_timestamp.

```
let all_data = {};
firestore()
.collection(collection_name)
.where('md', '>=', last_fetch_firestore_timestamp)
.onSnapshot({includeMetadataChanges: true}, (snap) => {
snap.docChanges().forEach((change) => {
    const doc = change.doc;
    const docName = doc.id;
    const docData = doc.data();
    if (!docData.md) {  // Can add doc.metadata.hasPendingWrites as an and condition to this
        docData['md'] = docData['lmd'];
    }
    all_data[docName] = docData;
});
```

The value if docData.md is not null when the internet is on as the server timestamp is created and it returns the document with the timestamp value. This works perfectly on IOS and Android.
But in offline mode, since the value is not set, the md value will be null. Hence I set it to lmd.
This works perfectly on IOS but fails Android. The document added/updated won't be part of snapshot itself. But if I remove the where clause (which is required to fetch delta documents), it works.

**Reply by AkshayAdiga1997** [reply]
Yeah. Need to check out if there's an issue in their libraries. But I think this is an important case to be solved.

**Reply by mikehardy** [reply]
It does sound important. The platforms should behave the same. But it sounds like it needs to be pursued with the firebase-android-sdk quickstart for verification. Very curious to hear results - it *could* be something we are doing as we wrap the APIs but things like this are usually deeper

**Reply by AkshayAdiga1997** [reply]
@mikehardy Hey! Any updates on this? I am not sure if I can work on the quickstart thing as it is completely native IOS/Android and I am not good at it. Any other way I can be of any help to solve this issue? I want this to be fixed for my upcoming startup. Pl help me out. Thanks.

**Reply by mikehardy** [reply]
Hi @AkshayAdiga1997 - sorry, I am not planning on spending any time on this. I'm waiting to hear your results

**Reply by AkshayAdiga1997** [reply]
@mikehardy Sorry if you don't mind can you explain what results are you expecting here? Will try my best to get it out soon.

**Reply by mikehardy** [reply]
@AkshayAdiga1997 

My prior comment: 

> This may be due to underlying platform differences - I strongly suspect it is actually.
If you've checked the actual code through here to make sure that the underlying javascript (then Obj-C and Java) API calls seem sensible, my best - though unfortunately difficult - suggestion is to make a quickstart for each platform where you verify behavior and confirm whether it's an upstream issue or not
> 
> https://github.com/firebase/quickstart-ios/tree/master/firestore
> https://github.com/firebase/quickstart-android/tree/master/firestore
> 
> I strongly recommend checking both of those repo's issues list (especially android since that appears to be the platform not meeting expectations?) to see if they have anything.
> 
> Also make absolutely sure you are on the most up to date react-native-firebase version so you have the updated native SDKs
